### PR TITLE
Deprecate HFCrossEntropy and Perplexity

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
This PR adds `DeprecationWarning`s to `HFCrossEntropy` and `Perplexity`, as the separation between these and `LanguageCrossEntropy` is confusing. To mitigate removing these, this PR also adds support for `Mapping` input to `LanguageCrossEntropy.update` and adds `LanguagePerplexity(LanguageCrossEntropy)`.

More context:
There is a slight difference between `LanguageCrossEntropy` and `HFCrossEntropy` due to how the loss is reduced. This creates confusion in the examples repo, which uses `LanguageCrossEntropy` and `Perplexity`. There is a possible small cost to this change, because `HFCrossEntropy` uses `output['loss']` (if available) from HF rather than recomputing the loss. `LanguageCrossEntropy` will always recompute the loss so that the reduction is consistent and `LanguagePerplexity` always matches `LanguageCrossEntropy`. The examples repo was always returning the logits from `forward` already, so this slight cost was already present in the examples repo.

# What issue(s) does this change relate to?
Closes [CO-1616](https://mosaicml.atlassian.net/browse/CO-1616)

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

[CO-1616]: https://mosaicml.atlassian.net/browse/CO-1616?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by dakinggg** [comment]
It looks like the difference is that `CrossEntropy` supports one hot targets, does not do the reshaping for you, and (with this PR) does not support `Mapping` input. On your comment about eliminating the need for `vocab_size`, yeah that seems good.

**Review by mvpatel2000 — APPROVED** [review]
Please don't merge until review from NLP person as well. LGTM for eng side

**Comment by dakinggg** [comment]
@abhi-mosaic @vchiley one of you mind taking a look to approve from the NLP side?

**Review by abhi-mosaic — APPROVED** [review]
Looks great! After this gets merged and released I will change our imports in `examples/[llm, bert]`, likely on 0.13 release

**Comment by mvpatel2000** [comment]
@dakinggg lets hold until after 12.1

**Comment by dakinggg** [comment]
@mvpatel2000 yeah, that was my plan
